### PR TITLE
Add a kernel for integer GEMM

### LINKED ISSUE #3941: matmul fails to handle overflows for uint64 properly
This code

```
import numpy as np
import cupy as cp

a = cp.random.random_integers(-2**63, 2**63 - 1, (1,)).astype(cp.uint64)
b = cp.random.random_integers(-2**63, 2**63 - 1, (1,)).astype(cp.uint64)
a_np = cp.asnumpy(a)
b_np = cp.asnumpy(b)
print(a * b, cp.matmul(a, b))
print(a_np * b_np, np.matmul(a_np, b_np))
```

can output, e.g.,

```
[11162831459571310510] 18446744073709551615
[11162831459571310510] 11162831459571310510
```

Versions
```
CuPy Version          : 7.8.0
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10020
CUDA Driver Version   : 10020
CUDA Runtime Version  : 10020
cuBLAS Version        : 10202
cuFFT Version         : 10102
cuRAND Version        : 10102
cuSOLVER Version      : (10, 3, 0)
cuSPARSE Version      : 10301
NVRTC Version         : (10, 2)
cuDNN Build Version   : 7605
cuDNN Version         : 7605
NCCL Build Version    : 2708
NCCL Runtime Version  : 2708
CUB Version           : None
cuTENSOR Version      : None
```

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #3941

We rely only on cublas for the matrix multiplication and we cast the inputs to float and then back to integral types again. This causes precision issues.

This also has an speedup of 45% for a matrix product with `dtype=numpy.int32` and`shape=(5000, 5000)`

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2743,64 +2764,306 @@ cpdef ndarray matmul(ndarray a, ndarray b, ndarray out=None):
...
+    blk_k=4
+    dim_xa=64
+    dim_ya=4
+    dim_xb=4
+    dim_yb=64
+    config = (('DIM_X', dim_x), ('DIM_Y', dim_y),
+              ('BLK_M', blk_m), ('BLK_N', blk_n), ('BLK_K', blk_k),
+              ('DIM_XA', dim_xa), ('DIM_YA', dim_ya),
+              ('DIM_XB', dim_xb), ('DIM_YB', dim_yb),
+              ('THR_M', blk_m // dim_x), ('THR_N', blk_n // dim_y))
```
Comment (emcastillo) [inline cupy/core/core.pyx:3011]: Change this

File: tests/cupy_tests/math_tests/test_matmul.py [hunk 2]
```diff
@@ -129,6 +129,19 @@ def test_cupy_matmul(self, xp, dtype1, dtype2):
         return xp.matmul(x1, x2)
 
 
+class TestMatmulOverflow(unittest.TestCase):
+
+    @testing.for_int_dtypes(name='dtype')
+    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-3)  # required for uint8
+    def test_overflow(self, xp, dtype):
```
Comment (emcastillo) [inline tests/cupy_tests/math_tests/test_matmul.py:136]: This test is verified to fail in current master

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -3012,11 +3012,13 @@ cpdef ndarray tensordot_core(
...
         for ace in _accelerator._routine_accelerators:
+            ret = _ndarray_init(ret_shape, dtype)
             # fast path using CUB or cuTENSOR
             if ace in (_accelerator.ACCELERATOR_CUB,
                        _accelerator.ACCELERATOR_CUTENSOR):
-                out = (a.ravel() * b.ravel()).sum(
-                    out=_manipulation._reshape(out, ()))
+                ret = (a.ravel() * b.ravel()).sum(
+                    out=_manipulation._reshape(ret, ()))
+                elementwise_copy(ret, out)
```
Comment (emcastillo) [inline cupy/core/core.pyx:3021]: Needs an extra array to avoid shape issue due to directly using the out one.

**Comment by emcastillo** [comment]
@takagi can you please review this before the release 😇

**Comment by takagi** [comment]
So sorry that I missed merging this before the code freeze...

**Comment by takagi** [comment]
LGTM!

**Pull request merged**